File: scripts/ingest/qcew.ts
import { parse } from 'csv-parse/sync';
import type { Observation } from '@dmv/shared';
import { fetchText } from '../lib/http.js';
import { log } from '../lib/log.js';
import { HttpError, IngestError } from '../lib/errors.js';
import { DMV_COUNTIES } from '../lib/counties.js';
import type { DataSource } from './DataSource.js';

const QCEW_BASE = 'https://data.bls.gov/cew/data/api';
const START_YEAR = 2015;
const QUARTERS = [1, 2, 3, 4] as const;
const CONCURRENCY = 4;

export interface QcewRow {
  area_fips: string;
  own_code: string;
  industry_code: string;
  agglvl_code: string;
  year: string;
  qtr: string;
  disclosure_code: string;
  month3_emplvl: string;
}

export function parseQcewCsv(csv: string): QcewRow[] {
  return parse(csv, { columns: true, skip_empty_lines: true }) as QcewRow[];
}

export function selectFederalCountyTotal(rows: QcewRow[]): QcewRow | null {
  for (const row of rows) {
    if (row.own_code === '1' && row.agglvl_code === '71' && row.industry_code === '10') {
      return row;
    }
  }
  return null;
}

export function quarterToObservedAt(year: number, qtr: 1 | 2 | 3 | 4): string {
  const suffix = qtr === 1 ? '03-01' : qtr === 2 ? '06-01' : qtr === 3 ? '09-01' : '12-01';
  return `${year}-${suffix}`;
}

export function rowToObservation(row: QcewRow, fips: string): Observation | null {
  if (row.disclosure_code === 'N') {
    log.warn({ fips, year: row.year, qtr: row.qtr }, 'qcew: suppressed; skipping');
    return null;
  }
  const emplvl = Number(row.month3_emplvl);
  if (!Number.isFinite(emplvl)) {
    log.warn({ fips, year: row.year, qtr: row.qtr, value: row.month3_emplvl }, 'qcew: non-finite value; skipping');
    return null;
  }
  const qtr = Number(row.qtr) as 1 | 2 | 3 | 4;
  const series = `qcew:${fips}:${row.year}Q${row.qtr}:own1:naics10`;
  return {
    source: 'qcew',
    series,
    fips,
    metric: 'federal_employment',
    observedAt: quarterToObservedAt(Number(row.year), qtr),
    value: emplvl,
    unit: 'count',
  };
}

export async function runWithConcurrency<T, R>(
  items: T[],
  limit: number,
  fn: (t: T) => Promise<R>,
): Promise<R[]> {
  const results: R[] = new Array(items.length);
  let nextIndex = 0;

  async function worker(): Promise<void> {
    while (true) {
      const i = nextIndex++;
      if (i >= items.length) return;
      const item = items[i] as T;
      results[i] = await fn(item);
    }
  }

  const workers: Promise<void>[] = [];
  const workerCount = Math.min(limit, items.length);
  for (let w = 0; w < workerCount; w++) {
    workers.push(worker());
  }
  await Promise.all(workers);
  return results;
}

interface FetchTask {
  fips: string;
  year: number;
  qtr: 1 | 2 | 3 | 4;
}

export class QcewSource implements DataSource {
  readonly name = 'qcew';
  readonly cadence = 'quarterly' as const;

  async fetch(): Promise<Observation[]> {
    const now = new Date();
    const currentYear = now.getUTCFullYear();
    const currentMonth = now.getUTCMonth() + 1;
    const currentQuarter = Math.ceil(currentMonth / 3) as 1 | 2 | 3 | 4;

    const tasks: FetchTask[] = [];
    for (let year = START_YEAR; year <= currentYear; year++) {
      for (const qtr of QUARTERS) {
        if (year === currentYear && qtr > currentQuarter) continue;
        for (const county of DMV_COUNTIES) {
          tasks.push({ fips: county.fips, year, qtr });
        }
      }
    }

    log.info({ count: tasks.length }, 'qcew: starting fetch');

    const results = await runWithConcurrency(tasks, CONCURRENCY, async (task) => {
      const url = `${QCEW_BASE}/${task.year}/${task.qtr}/area/${task.fips}.csv`;
      const label = `qcew:${task.fips}:${task.year}Q${task.qtr}`;
      try {
        const csv = await fetchText(url, { label });
        const rows = parseQcewCsv(csv);
        const row = selectFederalCountyTotal(rows);
        if (!row) {
          log.warn({ fips: task.fips, year: task.year, qtr: task.qtr }, 'qcew: federal county total row not found; skipping');
          return null;
        }
        return rowToObservation(row, task.fips);
      } catch (err) {
        if (err instanceof HttpError && err.status === 404) {
          log.warn({ fips: task.fips, year: task.year, qtr: task.qtr }, 'qcew: 404 (data not yet published); skipping');
          return null;
        }
        throw new IngestError(
          `qcew fetch failed for ${task.fips} ${task.year}Q${task.qtr}`,
          { source: 'qcew', fips: task.fips, url },
          err,
        );
      }
    });

    const observations: Observation[] = [];
    for (const r of results) {
      if (r !== null) observations.push(r);
    }

    log.info({ count: observations.length }, 'qcew: fetch complete');
    return observations;
  }
}

File: go/internal/ingest/qcew/qcew.go
// Package qcew ports scripts/ingest/qcew.ts. For every quarter from 2015 Q1
// through the current quarter, downloads the BLS QCEW per-county CSV, picks
// out the federal-county-total row (own_code=1, agglvl_code=71,
// industry_code=10), and emits one observation per (county, quarter).
package qcew

import (
	"context"
	"encoding/csv"
	"errors"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/meowmix1337/dmv_housing_hub/go/internal/counties"
	httpclient "github.com/meowmix1337/dmv_housing_hub/go/internal/http"
	pkglog "github.com/meowmix1337/dmv_housing_hub/go/internal/log"
	"github.com/meowmix1337/dmv_housing_hub/go/internal/types"
)

const (
	qcewBase    = "https://data.bls.gov/cew/data/api"
	startYear   = 2015
	concurrency = 4
)

// Row holds the few QCEW columns we care about.
type Row struct {
	AreaFIPS        string
	OwnCode         string
	IndustryCode    string
	AgglvlCode      string
	Year            string
	Qtr             string
	DisclosureCode  string
	Month3Emplvl    string
}

type task struct {
	fips string
	year int
	qtr  int
}

type Source struct {
	client   *httpclient.Client
	logger   *slog.Logger
	baseURL  string // override for tests
	now      func() time.Time
}

func New(client *httpclient.Client) *Source {
	return &Source{
		client:  client,
		logger:  pkglog.Default(),
		baseURL: qcewBase,
		now:     func() time.Time { return time.Now().UTC() },
	}
}

func (s *Source) Name() string           { return "qcew" }
func (s *Source) Cadence() types.Cadence { return types.CadenceQuarterly }

func (s *Source) Fetch(ctx context.Context) ([]types.Observation, error) {
	now := s.now()
	curYear := now.Year()
	curQtr := (int(now.Month())-1)/3 + 1

	cs := counties.All()
	tasks := make([]task, 0, len(cs)*12*4)
	for y := startYear; y <= curYear; y++ {
		for q := 1; q <= 4; q++ {
			if y == curYear && q > curQtr {
				continue
			}
			for _, c := range cs {
				tasks = append(tasks, task{fips: c.FIPS, year: y, qtr: q})
			}
		}
	}
	s.logger.Info("qcew: starting fetch", "count", len(tasks))

	results := make([]*types.Observation, len(tasks))
	var nextIdx int
	var mu sync.Mutex
	var wg sync.WaitGroup
	var fatalErr error
	var errOnce sync.Once

	workerN := concurrency
	if workerN > len(tasks) {
		workerN = len(tasks)
	}
	for w := 0; w < workerN; w++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for {
				mu.Lock()
				i := nextIdx
				nextIdx++
				mu.Unlock()
				if i >= len(tasks) {
					return
				}
				if ctx.Err() != nil {
					return
				}
				obs, err := s.fetchOne(ctx, tasks[i])
				if err != nil {
					errOnce.Do(func() { fatalErr = err })
					return
				}
				results[i] = obs
			}
		}()
	}
	wg.Wait()

	if fatalErr != nil {
		return nil, fatalErr
	}

	out := make([]types.Observation, 0, len(results))
	for _, r := range results {
		if r != nil {
			out = append(out, *r)
		}
	}
	s.logger.Info("qcew: fetch complete", "count", len(out))
	return out, nil
}

func (s *Source) fetchOne(ctx context.Context, t task) (*types.Observation, error) {
	url := fmt.Sprintf("%s/%d/%d/area/%s.csv", s.baseURL, t.year, t.qtr, t.fips)
	label := fmt.Sprintf("qcew:%s:%dQ%d", t.fips, t.year, t.qtr)

	body, err := s.client.GetText(ctx, url, label)
	if err != nil {
		var httpErr *httpclient.HTTPError
		if errors.As(err, &httpErr) && httpErr.Status == http.StatusNotFound {
			s.logger.Warn("qcew: 404 (data not yet published); skipping",
				"fips", t.fips, "year", t.year, "qtr", t.qtr)
			return nil, nil
		}
		return nil, fmt.Errorf("qcew fetch %s: %w", label, err)
	}

	rows, err := ParseCSV(body)
	if err != nil {
		return nil, fmt.Errorf("qcew parse %s: %w", label, err)
	}
	row := SelectFederalCountyTotal(rows)
	if row == nil {
		s.logger.Warn("qcew: federal county total row not found; skipping",
			"fips", t.fips, "year", t.year, "qtr", t.qtr)
		return nil, nil
	}
	obs := RowToObservation(*row, t.fips, s.logger)
	return obs, nil
}

// ParseCSV decodes the body into Rows. Tolerates extra columns; pulls fields by header name.
func ParseCSV(body string) ([]Row, error) {
	r := csv.NewReader(strings.NewReader(body))
	r.FieldsPerRecord = -1
	header, err := r.Read()
	if err != nil {
		return nil, fmt.Errorf("read header: %w", err)
	}
	idx := make(map[string]int, len(header))
	for i, h := range header {
		idx[h] = i
	}
	get := func(rec []string, name string) string {
		i, ok := idx[name]
		if !ok || i >= len(rec) {
			return ""
		}
		return rec[i]
	}

	out := make([]Row, 0, 200)
	for {
		rec, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, fmt.Errorf("read record: %w", err)
		}
		out = append(out, Row{
			AreaFIPS:       get(rec, "area_fips"),
			OwnCode:        get(rec, "own_code"),
			IndustryCode:   get(rec, "industry_code"),
			AgglvlCode:     get(rec, "agglvl_code"),
			Year:           get(rec, "year"),
			Qtr:            get(rec, "qtr"),
			DisclosureCode: get(rec, "disclosure_code"),
			Month3Emplvl:   get(rec, "month3_emplvl"),
		})
	}
	return out, nil
}

// SelectFederalCountyTotal picks the row matching own_code=1, agglvl_code=71, industry_code=10.
func SelectFederalCountyTotal(rows []Row) *Row {
	for i := range rows {
		r := &rows[i]
		if r.OwnCode == "1" && r.AgglvlCode == "71" && r.IndustryCode == "10" {
			return r
		}
	}
	return nil
}

// QuarterToObservedAt converts year+qtr to the ISO date used by TS (end-of-quarter month, day 01).
func QuarterToObservedAt(year, qtr int) string {
	var mm string
	switch qtr {
	case 1:
		mm = "03"
	case 2:
		mm = "06"
	case 3:
		mm = "09"
	case 4:
		mm = "12"
	default:
		return ""
	}
	return fmt.Sprintf("%d-%s-01", year, mm)
}

// RowToObservation converts a federal-county-total row to an Observation, or returns nil if
// the row is suppressed or non-numeric.
func RowToObservation(r Row, fips string, logger *slog.Logger) *types.Observation {
	if logger == nil {
		logger = pkglog.Default()
	}
	if r.DisclosureCode == "N" {
		logger.Warn("qcew: suppressed; skipping", "fips", fips, "year", r.Year, "qtr", r.Qtr)
		return nil
	}
	val, err := strconv.ParseFloat(r.Month3Emplvl, 64)
	if err != nil {
		logger.Warn("qcew: non-finite value; skipping",
			"fips", fips, "year", r.Year, "qtr", r.Qtr, "value", r.Month3Emplvl)
		return nil
	}
	year, err := strconv.Atoi(r.Year)
	if err != nil {
		return nil
	}
	qtr, err := strconv.Atoi(r.Qtr)
	if err != nil {
		return nil
	}
	return &types.Observation{
		Source:     "qcew",
		Series:     fmt.Sprintf("qcew:%s:%sQ%s:own1:naics10", fips, r.Year, r.Qtr),
		FIPS:       fips,
		Metric:     types.MetricFederalEmployment,
		ObservedAt: QuarterToObservedAt(year, qtr),
		Value:      val,
		Unit:       types.UnitCount,
	}
}
